pkg/sign/keys.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2024 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sign
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha256"
+	_ "crypto/sha512" // if user chooses SHA2-384 or SHA2-512 for hash
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+
+	protocommon "github.com/sigstore/protobuf-specs/gen/pb-go/common/v1"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+)
+
+type Keypair interface {
+	GetHashAlgorithm() protocommon.HashAlgorithm
+	GetHint() []byte
+	GetKeyAlgorithm() string
+	GetPublicKeyPem() (string, error)
+	SignDigest(digest []byte) ([]byte, error)
+}
+
+type EphemeralKeypairOptions struct {
+	// Optional hash algorithm to use to create digest of data provided to sign
+	HashAlgorithm protocommon.HashAlgorithm
+	// Optional hint of for signing key
+	Hint []byte
+}
+
+type EphemeralKeypair struct {
+	options    *EphemeralKeypairOptions
+	privateKey *ecdsa.PrivateKey
```

Comment:
I added a TODO - but I think changing this now is more trouble than it's worth, since `crypto.PrivateKey` doesn't include `Sign()`. Since it's unexported, it should be easy to change when we add support for more key types.